src/main/kotlin/com/hibob/academy/dao/PetDao.kt

```diff
@@ -0,0 +1,71 @@
+
+import com.hibob.academy.dao.*
+import jakarta.inject.Inject
+import org.jooq.DSLContext
+import org.jooq.Record
+import org.jooq.RecordMapper
+import org.springframework.stereotype.Component
+import java.sql.Date
+import java.util.*
+import kotlin.random.Random
+
+@Component
+
+class PetDao @Inject constructor(private val sql: DSLContext) {
+
+    private val pet = PetTable.instance
+
+    private val petMapper = RecordMapper<Record, PetData> { record ->
+        PetData(
+            record[pet.id],
+            record[pet.name],
+            PetType.convertStringToPetType(record[pet.type]),
+            record[pet.companyId],
+            Date(record[pet.dateOfArrival].time),
+            record[pet.ownerId]
+        )
+    }
+
+    private val petMapperWithoutType = RecordMapper<Record, PetDataWithoutType> { record ->
+        PetDataWithoutType(
+            record[pet.id],
+            record[pet.name],
+            record[pet.companyId],
+            Date(record[pet.dateOfArrival].time),
+            record[pet.ownerId]
+        )
+    }
+
+    fun getPetsByType(petType: PetType, companyId: Long): List<PetDataWithoutType> =
+        sql.select(pet.name, pet.companyId, pet.dateOfArrival, pet.ownerId)
+            .from(pet)
+            .where(pet.type.eq(PetType.convertPetTypeToPetString(petType))
+            .and(pet.companyId.eq(companyId)))
+            .fetch(petMapperWithoutType)
+
+    fun getAllPets(companyId: Long): List<PetData> =
+        sql.select(pet.name, pet.type, pet.companyId, pet.dateOfArrival, pet.ownerId)
+            .from(pet)
+            .where(pet.companyId.eq(companyId))
+            .fetch(petMapper)
+
+
+    fun createPet(id: Long, name: String, type: String, companyId: Long, dateOfArrival: Date, ownerId: Long) =
+        sql.insertInto(pet)
+            .set(pet.id, id)
+            .set(pet.name, name)
+            .set(pet.type, type)
+            .set(pet.companyId, companyId)
+            .set(pet.dateOfArrival, dateOfArrival)
+            .set(pet.ownerId, ownerId)
+            .onConflict(pet.companyId)
+            .doNothing()
+            .execute()
+
+    fun updatePetOwner(petData: PetData, ownerId: Long) =
+        sql.update(pet)
+            .set(pet.ownerId, ownerId)
+            .where(pet.id.eq(petData.id))
+            .and(pet.ownerId.isNull())
```

Comment:
What about the following scenario.
A pet was adopted and now someone else adopt it?
Do ownerId must be null? 
Moreover, add companyId check.

Reply:
In my implementation this scenario is not possible.
A pet can be adopted only if it wasn't adopted yet.
So in order to check if a pet was adopted or not I need to check it's owners id, if it's null it means the pet wasn't adopted yet.